### PR TITLE
Web support

### PR DESCRIPTION
**Pull request by movibe**
I added the tested theme support on windows 10, hope you enjoy it

### CODE REVIEW

**Comment by movibe** [comment]
chrome-capture

File: tsconfig.json [hunk 1]
```diff
@@ -13,7 +13,8 @@
     "moduleResolution": "node",
     "skipLibCheck": true,
     "lib": ["dom", "es2015", "es2016", "esnext"],
-    "jsx": "react-native"
+    "jsx": "react-native",
+    "noImplicitAny": false
```
Comment (brentvatne) [inline tsconfig.json]: why disable `noImplicitAny`?

File: src/polyfill.web.tsx [hunk 2]
```diff
@@ -1 +1,56 @@
...
+const isDarkQuery = window.matchMedia('(prefers-color-scheme: dark)');
+const isNoPreferenceQuery = window.matchMedia('(prefers-color-scheme: no-preference)');
+const isLightQuery = window.matchMedia('(prefers-color-scheme: light)');
+const isSupported = isDarkQuery.matches || isNoPreferenceQuery.matches || isLightQuery.matches;
+
+const initialMode: Mode = isSupported
+  ? window.matchMedia('(prefers-color-scheme: dark)').matches
+    ? 'dark'
+    : 'light'
+  : 'light';
```
Comment (brentvatne) [inline src/polyfill.web.tsx]: it should default to `'no-preference'` if not supported, see: https://snack.expo.io/Hy-xEWeYB

**Review by brentvatne — CHANGES_REQUESTED** [review]

File: src/polyfill.web.tsx (hunk 2 above)
Comment (movibe) [reply src/polyfill.web.tsx]: done, can you check?

File: tsconfig.json (hunk 1 above)
Comment (movibe) [reply tsconfig.json]: done, can you check?

**Comment by FiberJW** [comment]
fixes https://github.com/expo/expo/issues/5463

File: example/App.tsx [hunk 3]
```diff
@@ -2,15 +2,26 @@ import * as React from 'react';
 import { StyleSheet, Text, View, Button } from 'react-native';
 
 import { Appearance, AppearanceProvider, useColorScheme } from '..';
+import { AppearancePreferences } from '../src/Appearance.types';
 
 export default () => {
   let colorScheme = useColorScheme();
+  const [Dark, setDark] = React.useState(colorScheme);
```
Comment (EvanBacon) [inline example/App.tsx]:
```suggestion
  const [isDarkModeEnabled, setDarkModeEnabled] = React.useState(colorScheme);
```

File: src/polyfill.web.tsx [hunk 4]
```diff
@@ -1 +1,56 @@
...
+
+  static set(_preferences: AppearancePreferences): void {}
+
+  static addChangeListener(_listener: AppearanceListener): EventSubscription {
+    isDarkQuery.addListener(e => {
+      const colorScheme = e.matches ? 'dark' : 'light';
+      _listener({ colorScheme });
+      this.set({ colorScheme });
+    });
+    return { remove: () => removeListener };
```
Comment (EvanBacon) [inline src/polyfill.web.tsx]: good work

File: src/polyfill.web.tsx [hunk 5]
```diff
@@ -1 +1,56 @@
...
+  }
+
+  /**
+   * Unused: some people might expect to remove the listener like this, but they shouldn't.
+   */
+  static removeChangeListener(_listener: AppearanceListener): void {
```
Comment (EvanBacon) [inline src/polyfill.web.tsx]: Possibly add a warning that is wrapped in `__DEV__` to inform users how to properly remove the listener

File: src/polyfill.web.tsx [hunk 6]
```diff
@@ -1 +1,56 @@
...
+    removeListener();
+  }
+}
+
+export const AppearanceProvider = (props: any) => <View style={{ flex: 1 }} {...props} />;
```
Comment (EvanBacon) [inline src/polyfill.web.tsx]: Try using a `React.forwardRef` here so we can cut down on `div`s and avoid adding unexpected styling behavior. I can see this becoming a problem if someone is trying to enable/disable body scrolling.

File: src/polyfill.web.tsx [hunk 7]
```diff
@@ -1 +1,56 @@
...
+
+export class Appearance {
+  static getColorScheme(): ColorSchemeName {
+    return initialMode;
+  }
+
+  static set(_preferences: AppearancePreferences): void {}
+
+  static addChangeListener(_listener: AppearanceListener): EventSubscription {
+    isDarkQuery.addListener(e => {
```
Comment (EvanBacon) [inline src/polyfill.web.tsx]:
```suggestion
    isDarkQuery.addListener(({ matches }) => {
```

File: src/polyfill.web.tsx [hunk 8]
```diff
@@ -1 +1,56 @@
-export * from './mock';
\ No newline at end of file
+import * as React from 'react';
+import { View } from 'react-native';
+import { AppearancePreferences, ColorSchemeName, AppearanceListener } from './Appearance.types';
+
+type Mode = 'light' | 'dark' | 'no-preference';
+
+const isNoPreferenceQuery = window.matchMedia('(prefers-color-scheme: no-preference)');
+const isDarkQuery = window.matchMedia('(prefers-color-scheme: dark)');
```
Comment (EvanBacon) [inline src/polyfill.web.tsx]: You should only need one of these to get supported.

File: src/polyfill.web.tsx [hunk 9]
```diff
@@ -1 +1,56 @@
...
+const isLightQuery = window.matchMedia('(prefers-color-scheme: light)');
+const isSupported = isDarkQuery.matches || isNoPreferenceQuery.matches || isLightQuery.matches;
+
+const initialMode: Mode = isSupported
+  ? window.matchMedia('(prefers-color-scheme: dark)').matches
```
Comment (EvanBacon) [inline src/polyfill.web.tsx]:
```suggestion
  ? isDarkQuery.matches
```

File: src/polyfill.web.tsx [hunk 10]
```diff
@@ -1 +1,56 @@
-export * from './mock';
\ No newline at end of file
+import * as React from 'react';
+import { View } from 'react-native';
```
Comment (EvanBacon) [inline src/polyfill.web.tsx]: Attempt to make this agnostic of RN https://github.com/expo/react-native-appearance/pull/7/files#r334293600

File: src/polyfill.web.tsx [hunk 11]
```diff
@@ -1 +1,56 @@
...
+import { AppearancePreferences, ColorSchemeName, AppearanceListener } from './Appearance.types';
+
+type Mode = 'light' | 'dark' | 'no-preference';
+
+const isNoPreferenceQuery = window.matchMedia('(prefers-color-scheme: no-preference)');
```
Comment (EvanBacon) [inline src/polyfill.web.tsx]: https://www.w3schools.com/jsref/met_win_matchmedia.asp

```js
const query = window.matchMedia
  ? window.matchMedia(`(prefers-color-scheme: no-preference)`)
  : {};
```

File: example/App.tsx [hunk 12]
```diff
@@ -2,15 +2,26 @@ import * as React from 'react';
...
+
+  React.useEffect(() => {
+    let subscription = Appearance.addChangeListener((preferences: AppearancePreferences) => {
```
Comment (EvanBacon) [inline example/App.tsx]:
```suggestion
    const subscription = Appearance.addChangeListener((preferences: AppearancePreferences) => {
```

File: example/App.tsx [hunk 13]
```diff
@@ -2,15 +2,26 @@ import * as React from 'react';
...
+      setDark(preferences.colorScheme);
+    });
+    return () => {
+      subscription.remove();
+    };
```
Comment (EvanBacon) [inline example/App.tsx]: xit

```suggestion
 return () => subscription.remove();
```

File: example/App.tsx [hunk 14]
```diff
@@ -2,15 +2,26 @@ import * as React from 'react';
 import { StyleSheet, Text, View, Button } from 'react-native';
 
 import { Appearance, AppearanceProvider, useColorScheme } from '..';
+import { AppearancePreferences } from '../src/Appearance.types';
 
 export default () => {
   let colorScheme = useColorScheme();
```
Comment (EvanBacon) [inline example/App.tsx]: Please make this update with `useEffect` internally.

**Comment by EvanBacon** [comment]
This [type of import](https://github.com/expo/react-native-appearance/blob/9ce8f4504500471908dac5b4538e5b23b0cda449/src/index.tsx#L4-L6) will break tree-shaking on web. This is something that Expo for web has prohibited as it's very difficult to debug.

**Comment by EvanBacon** [comment]
I'll work on implementing some of these changes and squashing them onto the branch.

**Comment by movibe** [comment]
@EvanBacon can i help in any way?

**Comment by brentvatne** [comment]
landed this in https://github.com/expo/react-native-appearance/commit/1e31fbd50e8f867dfccceb645d5073001b485242!

**Pull request closed without merging**